### PR TITLE
Fix building Flatcar images for Flatcar stable releases after 3139.2.0

### LINKED ISSUE #861: image_version is not being passed to resulting CAPI SIG images on Azure
**What steps did you take and what happened:**
Discovered while investigating https://github.com/flatcar-linux/Flatcar/issues/719 and https://github.com/kubernetes-sigs/image-builder/issues/859. Turns out `image_version` parameter is only passed to the first builder in https://github.com/kubernetes-sigs/image-builder/blob/a09b089b1b344d75275cf741cafdbd877050f660/images/capi/packer/azure/packer.json#L2, which makes resulting image always use latest version of SKU, even if different one is specified.

**What did you expect to happen:**

Specified image version to be used as a base image.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

image-builder version a09b089b1

CC @jepio

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### LINKED ISSUE #862: Building Flatcar images fails due to os_family Ansible mismatch
**What steps did you take and what happened:**

Because of https://github.com/kubernetes-sigs/image-builder/issues/861 and the regression in recent Flatcar stable release https://github.com/flatcar-linux/Flatcar/issues/719, building SIG Flatcar images is right now failing.

**What did you expect to happen:**

Ci to succeed.

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by invidian**
What this PR does / why we need it:

Fixes for building Flatcar Azure SIG images related to recent regression. See commit messages for more details.

Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):

Closes #862
Closes #861

### CODE REVIEW

**Comment by invidian** [comment]
Updated to include more generic fix. Let's see if that fixes the CI.

File: images/capi/ansible/roles/python/tasks/main.yml [hunk 1]
```diff
@@ -17,5 +17,4 @@
   register: distrib_id
 
 - include_tasks: flatcar.yml
-  when: distrib_id.stdout_lines[0] is search("Flatcar")
-
+  when: ansible_os_family in ["Flatcar", "Flatcar Container Linux by Kinvolk"]
```
Comment (invidian) [inline images/capi/ansible/roles/python/tasks/main.yml:20]: We still end up with couple of `ansible_os_family in ["Flatcar", "Flatcar Container Linux by Kinvolk"]` and many previously used `ansible_os_family == "Flatcar"`.

Ideally, we would be able to unify it and this will happen, when Ansible is fixed and updates, however, it may take a while.

Alternative path would be to use `set_fact` at some higher level in execution hierarchy, e.g. in the playbook before roles are applied. It could override either `ansible_os_family` or introduce some custom fact like `is_flatcar`, which should be easier to use, however inconsistent with other existing uses of `ansible_os_family`.

To summarize, I think proposed set of changes is not ideal, it's "long-term temporary", but it should do the job and it minimizes the changes needed now and in the future.

**Comment by invidian** [comment]
Ubuntu error on CI seems to be a fluke, but Flatcar is something I need to look into.

**Comment by Ankitasw** [comment]
hi @invidian, any more changes left in this PR, as this change works fine while building CAPA AMI images for flatcar OS. Any idea for the timeline of new release with this fix?

File: images/capi/ansible/python.yml [hunk 2]
```diff
@@ -13,7 +13,6 @@
 # limitations under the License.
 ---
 - hosts: all
-  gather_facts: no
```
Comment (invidian) [inline images/capi/ansible/python.yml:16]: > hi @invidian, any more changes left in this PR, as this change works fine while building CAPA AMI images for flatcar OS. Any idea for the timeline of new release with this fix?

So this change does not work without Python installed and this playbook does exactly that, so I have to re-evaluate how to make Flatcar detection here, as it seems it cannot be consistent with other roles. `ansible_os_family` doesn't seem available, while `distrib_id` definitely is and `os_release` may probably be available as well.

**Comment by invidian** [comment]
The CI is passing now, but I'd appreciate more opinions on introduced changes. @jepio or @johananl maybe one of you could have a look please?

File: images/capi/ansible/roles/setup/tasks/flatcar.yml [hunk 3]
```diff
@@ -21,7 +21,7 @@
   environment: {}
 
 - include_tasks: bootstrap-flatcar.yml
-  when: '"Flatcar" in os_release.stdout'
+  when: ansible_os_family in ["Flatcar", "Flatcar Container Linux by Kinvolk"]
```
Comment (jepio) [inline images/capi/ansible/roles/setup/tasks/flatcar.yml:24]: This check could be removed since we are already in `flatcar.yml`. That also removes the need to check /etc/os-release here.

File: images/capi/ansible/roles/setup/tasks/main.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@
   when: ansible_os_family == "Debian"
 
 - import_tasks: flatcar.yml
-  when: ansible_os_family == "Flatcar"
+  when: ansible_os_family in ["Flatcar", "Flatcar Container Linux by Kinvolk"]
```
Comment (jepio) [inline images/capi/ansible/roles/setup/tasks/main.yml:19]: This could use the comment about flatcar.yml being the one that overrides `ansible_os_family` to make it uniform everywhere else.

File: images/capi/ansible/python.yml (hunk 2 above)
Comment (jepio) [reply images/capi/ansible/python.yml:16]: Interesting thing is this (python.yml) playbook is only ever used by the azure builder. All other builders directly call `bootstrap-flatcar.sh` as a script.

File: images/capi/ansible/roles/setup/tasks/flatcar.yml (hunk 3 above)
Comment (invidian) [reply images/capi/ansible/roles/setup/tasks/flatcar.yml]: Done, thanks!

**Comment by invidian** [comment]
CI failures seems random and unrelated to the PR content :cry: a

File: images/capi/ansible/python.yml [hunk 5]
```diff
@@ -13,6 +13,8 @@
 # limitations under the License.
 ---
 - hosts: all
+  # Gathering facts require Python to be available, so it's a chicken egg
```
Comment (johananl) [inline images/capi/ansible/python.yml:16]:
```suggestion
  # Gathering facts requires Python to be available, so it's a chicken and egg
```

File: images/capi/ansible/python.yml [hunk 6]
```diff
@@ -13,6 +13,8 @@
...
+  # problem, as this playbook installs Python.
```
Comment (johananl) [inline images/capi/ansible/python.yml:17]:
```suggestion
  # problem as this playbook installs Python.
```

File: images/capi/ansible/roles/python/tasks/main.yml [hunk 7]
```diff
@@ -17,5 +17,6 @@
   register: distrib_id
 
 - include_tasks: flatcar.yml
+  # We can't use ansible_os_family fact here for consistency, as facts gathering
+  # is disabled in playbook including this role. See playbook for more details.
```
Comment (johananl) [inline images/capi/ansible/roles/python/tasks/main.yml:21]:
```suggestion
  # is disabled in the playbook which includes this role. See playbook for more details.
```

Resolve ambiguity.

File: images/capi/ansible/roles/setup/tasks/flatcar.yml [hunk 8]
```diff
@@ -21,7 +21,6 @@
   environment: {}
 
 - include_tasks: bootstrap-flatcar.yml
-  when: '"Flatcar" in os_release.stdout'
```
Comment (johananl) [inline images/capi/ansible/roles/setup/tasks/flatcar.yml:24]: Can't we remove the `Fetch /etc/os-release` task, now that we don't care about its output?

File: images/capi/ansible/roles/setup/tasks/main.yml [hunk 9]
```diff
@@ -16,7 +16,10 @@
   when: ansible_os_family == "Debian"
 
 - import_tasks: flatcar.yml
-  when: ansible_os_family == "Flatcar"
+  # This task will override ansible_os_family to "Flatcar" as a workaround for
```
Comment (johananl) [inline images/capi/ansible/roles/setup/tasks/main.yml:19]:
```suggestion
  # This task overrides ansible_os_family to "Flatcar" as a workaround for
```

Avoid future tense in docs.

File: images/capi/ansible/roles/setup/tasks/main.yml [hunk 10]
```diff
@@ -16,7 +16,10 @@
...
+  # regression between Flatcar and Ansible, so rest of the code can use just
+  # "Flatcar" for comparison, which is the desired value.
```
Comment (johananl) [inline images/capi/ansible/roles/setup/tasks/main.yml:21]:
```suggestion
  # "Flatcar" for comparison, which is the correct value.
```

**Review by johananl — APPROVED** [review]
This seems to work well on my testing. Nice job overriding the fact rather than modifying all the `when` cases.

Added some non-blocking nits.

File: images/capi/ansible/roles/setup/tasks/flatcar.yml (hunk 8 above)
Comment (invidian) [reply images/capi/ansible/roles/setup/tasks/flatcar.yml:24]: Good point! Removed :)

**Comment by codenrhoden** [comment]
I like the way this looks now, nice work incorporating all the feedback, @invidian. 

It appears this fixes an outstanding issue with Flatcar in the Azure sigs test, so it will be nice to have that test stabilize. I will keep an eye out if any other feedback comes in, but I'd like to see this merge today.
